spatialmath/mesh.go

```diff
@@ -29,10 +40,279 @@ func (m *Mesh) Triangles() []*Triangle {
 }
 
 // Transform transforms the mesh. As triangles are in the mesh's frame, they are unchanged.
-func (m *Mesh) Transform(pose Pose) *Mesh {
+func (m *Mesh) Transform(pose Pose) Geometry {
 	// Triangle points are in frame of mesh, like the corners of a box, so no need to transform them
 	return &Mesh{
 		pose:      Compose(pose, m.pose),
 		triangles: m.triangles,
+		label:     m.label,
+	}
+}
+
+// CollidesWith checks if the given mesh collides with the given geometry and returns true if it does.
+func (m *Mesh) CollidesWith(g Geometry, collisionBufferMM float64) (bool, error) {
+	switch other := g.(type) {
+	case *box:
+		// Mesh-ifying the box misses the case where the box encompasses a mesh triangle without its surface intersecting a triangle.
+		encompassed := m.boxIntersectsVertex(other)
+		if encompassed {
+			return true, nil
+		}
+		// Convert box to mesh and check triangle collisions
+		return m.collidesWithMesh(other.toMesh(), collisionBufferMM), nil
```

Comment:
I dont understand why both this and the above function are necessary.

Reply:
There are two ways in which a box and mesh can collide:
- The box cleanly encompasses at least one triangle of the mesh
- The box surface is penetrated by at least one triangle

The two checks here test the first, then the second possibility.